src/link/Wasm.zig

```diff
@@ -651,6 +651,124 @@ fn resolveSymbolsInArchives(wasm: *Wasm) !void {
+
+    // when false, we fail linking. We only verify this after a loop to catch all invalid features.
+    var valid_feature_set = true;
+
+    // When the user has given an explicit list of features to enable,
+    // we extract them and insert each into the 'allowed' list.
+    if (!infer) {
+        try allowed.ensureUnusedCapacity(std.Target.wasm.all_features.len);
+        // std.builtin.Type.EnumField
+        inline for (@typeInfo(std.Target.wasm.Feature).Enum.fields) |feature_field| {
+            if (cpu_features.isEnabled(feature_field.value)) {
+                allowed.putAssumeCapacityNoClobber(@intToEnum(types.Feature.Tag, feature_field.value), {});
+            }
+        }
+    }
+
+    // extract all the used, disallowed and required features from each
+    // linked object file so we can test them.
+    for (wasm.objects.items) |object| {
+        for (object.features) |feature| {
+            switch (feature.prefix) {
+                .used => {
+                    const gop = try used.getOrPut(feature.tag);
+                    if (!gop.found_existing) {
+                        gop.value_ptr.* = object.name;
+                    }
+                },
+                .disallowed => {
+                    const gop = try disallowed.getOrPut(feature.tag);
+                    if (!gop.found_existing) {
+                        gop.value_ptr.* = object.name;
+                    }
+                },
+                .required => {
+                    const gop = try required.getOrPut(feature.tag);
+                    if (!gop.found_existing) {
+                        gop.value_ptr.* = object.name;
+                    }
+                    const used_gop = try used.getOrPut(feature.tag);
+                    if (!used_gop.found_existing) {
+                        used_gop.value_ptr.* = object.name;
+                    }
+                },
+            }
+        }
+    }
+
+    // when we infer the features, we allow each feature found in the 'used' set
+    // and insert it into the 'allowed' set. When features are not inferred,
+    // we validate that a used feature is allowed.
+    if (infer) try allowed.ensureUnusedCapacity(@intCast(u32, used.count()));
+    for (used.keys()) |used_feature, used_index| {
+        if (infer) {
+            allowed.putAssumeCapacityNoClobber(used_feature, {});
+        } else if (!allowed.contains(used_feature)) {
+            log.err("feature '{s}' not allowed, but used by linked object", .{@tagName(used_feature)});
+            log.err("  defined in '{s}'", .{used.values()[used_index]});
+            valid_feature_set = false;
+        }
+    }
+
+    if (!valid_feature_set) {
+        return error.InvalidFeatureSet;
+    }
+
+    // For each linked object, validate the required and disallowed features
+    for (wasm.objects.items) |object| {
+        var object_used_features = std.AutoHashMap(types.Feature.Tag, void).init(arena);
```

Comment:
I don't know how dynamic those `Feature`s are, but if they have to be known at compile time, and thus the set is finite, and if it is relatively small, you could optimise out any use of a hashmap for simple hard-coded (and always constant) feature-to-index lookup in a fixed-size array where each element could be a simple `u1`, so a `StaticBitSet` or similar.

Reply:
You could even store a set of such arrays/bitsets per each `Object` and provide accessor functions which would decode from a set bit into a `bool` or what have you if the object defines/disallows/allows/etc such a feature.

Reply:
This is indeed possible as I already applied that here: https://github.com/ziglang/zig/blob/3f70f670785233fed5436322731eb0ec68e9ab81/src/link/Wasm.zig#L657
I'll change all the hashmaps to apply the same logic. I think using a bitset to store all the features and its prefix is an excellent idea, especially since an array of bool already has 7 bits wasted that could be used to store whether it's used, required, or disallowed. Though, I'd rather make that change once I refactor `Object` as a whole. For now, I'll change the hashmaps in the `validateFeatures` function to separate arrays. Although this does mean that I'll eventually have to refactor that function also, I'd prefer to get more feature-complete first so I have a standard to benchmark against when trying to improve the linker's performance as a whole.